From 9779e130daf09a9638cb91a4f65ed1fd3f66bee3 Mon Sep 17 00:00:00 2001
From: Murala S Swamy <ssmurala@habana.ai>
Date: Tue, 14 Jan 2025 15:22:31 +0200
Subject: [PATCH] [FSW-12194] fs1: handle jtag security config

JTAG master interface range registers are
configured to open all addresses and protection
bits are already open by default when jtag
enabled. When JTAG disabled, this patch will
disable all range register sets in master
interface.

Change-Id: I90ef85f4cff1c338a0ccb1581701e08bbe5cc996
---

diff --git a/arch/fs1/arc/dev_security.c b/arch/fs1/arc/dev_security.c
index 193ad28..92a4ac6 100644
--- a/arch/fs1/arc/dev_security.c
+++ b/arch/fs1/arc/dev_security.c
@@ -92,6 +92,10 @@
 		mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_I2C_S_MSTR_IF_LBW_RR_BASE,
 		mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_I2C_S_MSTR_IF_LBW_AXI_OVRD_AXUSER_BASE,
 	},
+	[SECURITY_INITIATOR_JTAG_MSTR] = {
+		mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_JT_MSTR_IF_LBW_RR_BASE,
+		mmIO_0_CFG_HALF_0_PSOC_PSOC_CFG_JT_MSTR_IF_LBW_AXI_OVRD_AXUSER_BASE,
+	},
 };
 
 struct security_rr_ctx lbw_rr_ctx[SEC_ACCESS_CTRL_MAX] = {
@@ -121,6 +125,13 @@
 		.data_min = CONV_ARC_REG_TO_GEN_MEM(BOOT_ROM_REC_RR_GEN_MEM_MIN),
 		.data_max = CONV_ARC_REG_TO_GEN_MEM(BOOT_ROM_REC_RR_GEN_MEM_MAX),
 	},
+};
+
+struct security_rr_ctx lbw_rr_jtag_ctx = {
+	.rr_idx = LBW_RR_IDX_0,
+	.acc_type = SEC_ACCESS_RD_WR,
+	.data_min = 0x0,
+	.data_max = 0x3FFFFFFF,
 };
 
 /* Assign memory to PB_REG_BITS */
@@ -420,6 +431,16 @@
 	regmap_writel(rr_en_val, initr_data.base_addr + mmRANGE_REG_LBW_RANGE_EN_0 + rr_dist);
 }
 
+static void security_rr_lbw_config_dis(struct security_initiator_ctx initr_data,
+	uint32_t rr_idx)
+{
+	uint32_t rr_dist, dis_val = 0;
+
+	rr_dist = rr_idx * sizeof(uint32_t);
+
+	regmap_writel(dis_val, initr_data.base_addr + mmRANGE_REG_LBW_RANGE_EN_0 + rr_dist);
+}
+
 void security_rr_recovery_config(bool rr_en)
 {
 	uint8_t i;
@@ -438,6 +459,20 @@
 	}
 }
 
+static void security_rr_jtag_config(bool enable)
+{
+	uint32_t rr_idx;
+
+	if (enable) {
+		security_rr_lbw_config_en(initiator_data[SECURITY_INITIATOR_JTAG_MSTR],
+			&lbw_rr_jtag_ctx);
+	} else {
+		for (rr_idx = LBW_RR_IDX_0; rr_idx <= LBW_RR_IDX_15; rr_idx++)
+			security_rr_lbw_config_dis(initiator_data[SECURITY_INITIATOR_JTAG_MSTR],
+				rr_idx);
+	}
+}
+
 void security_handle_jtag_if(void)
 {
 	uint32_t val = regmap_readl(IO0_H0_PSOC_BOOT(mmPSOC_BOOT_CONF_JTAG_DISABLE));
@@ -446,23 +481,19 @@
 		/* Enable JTAG */
 		val &= ~(PSOC_BOOT_CONF_JTAG_DISABLE_VAL_M);
 
+		security_rr_jtag_config(true);
+
 		/*
-		 * TODO: RR
-		 * JTAG.MSTR_IF RANGE_EN = 1
-		 * WR_RANGE_SEC & RD_RANGE_SEC = JTAG_SAI
-		 * RANGE_MIN_ADDR & RANGE_MAX_ADDR = 0 to 0xFFFFFFFC
-		 * TODO: PB
+		 * Protection Bits:
 		 * Already all PBs are opened for RW in all pages for all SAI levels.
 		 */
 	} else {
 		val |= PSOC_BOOT_CONF_JTAG_DISABLE_VAL_M;
 
+		security_rr_jtag_config(false);
+
 		/*
-		 * TODO: RR
-		 * JTAG.MSTR_IF RANGE_EN = 0
-		 * WR_RANGE_SEC & RD_RANGE_SEC = 0
-		 * RANGE_MIN_ADDR & RANGE_MAX_ADDR = 0
-		 * TODO: PB
+		 * Protection Bits:
 		 * Blocked from RR level, so no need to change PB.
 		 */
 	}
diff --git a/arch/fs1/arc/dev_security.h b/arch/fs1/arc/dev_security.h
index 3cc6d93..51f3d6d 100644
--- a/arch/fs1/arc/dev_security.h
+++ b/arch/fs1/arc/dev_security.h
@@ -115,6 +115,7 @@
 enum security_initiators {
 	SECURITY_INITIATOR_PCIE,
 	SECURITY_INITIATOR_I2C_S_MSTR,
+	SECURITY_INITIATOR_JTAG_MSTR,
 	SECURITY_INITIATOR_MAX,
 };
 
